rfcs/20240325-reorganize-gpu/README.md

```diff
@@ -0,0 +1,115 @@
+enabled individually
+
+```bash
+├── cpu/ # CPU code
+├── sycl/ # A common place for basic CPU and GPU SYCL abstractions, e.g. `sycl_cpu_engine_t`, `sycl_gpu_engine_t`, etc.
+└── gpu/ # GPU code. Basic GPU code resides in the GPU directory directly, e.g. `gpu_engine_t`.
+    ├── intel/ # Everything that is related to Intel
+    │   ├── compute/ # Compute layer abstractions
+    │   ├── ocl/ # OpenCL kernels and abstractions (very similar to the current state of `gpu/ocl` directory)
+    │   ├── jit/ # JIT kernels/generators and abstractions (very similar to the current state of `gpu/jit` directory)
+    │   ├── sycl/ # Intel related SYCL functionality/abstractions (e.g. SYCL kernels that use eSIMD, Intel specific extensions, etc)
+    │   └── ...
+    ├── nvidia/ # Everything that is related to NVIDIA
+    │   ├── sycl # NVIDIA related functionality/abstractions (cuDNN/cuBLAS based and CUDA specific SYCL kernels)
+    │   └── ...
+    ├── amd/ # Everything that is related to AMD
+    │   ├── sycl/ # AMD related functionality/abstractions (MIOpen/rocBLAS based and HIP specific SYCL kernels)
+    │   └── ...
+    └── generic/ # Everything that is related to generic kernels
+        └── sycl/ # generic SYCL kernels and related abstractions
+
+```
+
+### Affected Basic Abstractions
+
+The new schema will require moving a lot of parts of the library around
+and while most of the changes are probably just an implementation detail
+there are a few major changes that have to be described in this RFC.
+
+#### Engine
+
+Engine is the most heavily affected abstraction by the changes.
+There is currently no `gpu_engine_t` class as `compute_engine_t` is used as
+an Intel centric alternative of it. In order to have a common basic
+abstraction at the GPU level the `gpu_engine_t` class will be introduced.
+
+The current inheritance chains for SYCL and OpenCL GPU engines are
+the following:
+* SYCL Intel and generic: `engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_gpu_engine_t`
+* SYCL NVIDIA: `engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_cuda_engine_t`
+* SYCL AMD: `engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_hip_engine_t`
+* OpenCL (only Intel): `engine_t` -> `compute_engine_t` -> `ocl_gpu_engine_t`
+
+With the new schema they will be changed to:
+* SYCL Generic: `engine_t` -> `gpu_engine_t` -> `gpu_generic_engine_t` -> `sycl_engine_base_t` -> `sycl_gpu_generic_engine_t`
+* SYCL Intel: `engine_t` -> `gpu_engine_t` -> `gpu_intel_engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_gpu_intel_engine_t`
+* SYCL NVIDIA: `engine_t` -> `gpu_engine_t` -> `gpu_nvidia_engine_t` -> `sycl_engine_base_t` -> `sycl_gpu_nvidia_engine_t`
+* SYCL AMD: `engine_t` -> `gpu_engine_t` -> `gpu_amd_engine_t` -> `sycl_engine_base_t` -> `sycl_gpu_amd_engine_t`
+* OpenCL: `engine_t` -> `gpu_engine_t` ->  `gpu_intel_engine_t` -> `compute_engine_t` -> `ocl_gpu_engine_t`
```

Comment:
This feels to me like we are missing some abstractions of the runtime which is causing unnecessary inheritance. A 
 lot of the runtime abstraction is currently coming from the `compute_*_t` code.  What if we broke that out into its own thing and had something more like:
```
// runtime_t encapsulates runtime kernel object creation, and runtime support info, at its core, i believe this should be  basically be represented by a  runtime_device_id, and runtime_device_context.
ocl_runtime_t 
sycl_runtime_t

// Engines implementations contain a `runtime_t` object used for implementing the engine_t interface along
 with any <vendor>_devic_info_t that may not be available from the the runtime.
SYCL NVIDIA: engine_t -> nvidia_sycl_engine_t
SYCL AMD:  engine_t -> amd_sycl_engine_t
SYCL INTEL:  engine_t -> intel_sycl_engine_t
OCL INTEL: engine_t -> intel_ocl_engine_t
```
Then rather than have a `generic` folder, we have some runtime folders `opencl` and `sycl` where the runtime implementation and generic kernel implementations are stored.

Reply:
If I get it correctly, the idea sounds very similar to what I described in [my comment](https://github.com/oneapi-src/oneDNN/pull/1840#discussion_r1542411980) above (the `sycl_engine_base_t` part). I think we will probably end up having something like that anyway because as I said in the comment duplicating `sycl_engine_base_t` doesn't look right. I haven't thought about OpenCL though as it doesn't have the engine base class problem but I like the general idea and I'm currently figuring out what the design should look like.

Also, we need to keep in mind that for SYCL runtime we need to be able to support the following scenarios: only generic kernels enabled, NVIDIA + generic kernels enabled and AMD + generic kernels enabled. That's why we should not tie the generic kernels to any particular vendor.

Reply:
@rjoursler, I updated the RFC, please take a look.